rest/diagnostic_api.go

```diff
@@ -0,0 +1,158 @@
+//  Copyright 2013-Present Couchbase, Inc.
+//
+//  Use of this software is governed by the Business Source License included
+//  in the file licenses/BSL-Couchbase.txt.  As of the Change Date specified
+//  in that file, in accordance with the Business Source License, use of this
+//  software will be governed by the Apache License, Version 2.0, included in
+//  the file licenses/APL2.txt.
+
+package rest
+
+import (
+	"github.com/couchbase/sync_gateway/auth"
+	"github.com/couchbase/sync_gateway/base"
+	"github.com/gorilla/mux"
+)
+
+// each field in this response is a map of collection names to a map of channel names to grant history info
+type getAllChannelsResponse struct {
+	AdminGrants       map[string]map[string]auth.GrantHistory            `json:"admin_grants,omitempty"`
+	DynamicGrants     map[string]map[string]auth.GrantHistory            `json:"dynamic_grants,omitempty"`
+	AdminRoleGrants   map[string]map[string]map[string]auth.GrantHistory `json:"admin_role_grants,omitempty"`
+	DynamicRoleGrants map[string]map[string]map[string]auth.GrantHistory `json:"dynamic_role_grants,omitempty"`
+	RoleHistoryGrants map[string]map[string]map[string]auth.GrantHistory `json:"role_history_grants,omitempty"`
```

Comment:
What are `role_history_grants` and why are they different than `dynamic_role_grants` or `admin_role_grants`? Are they just ones a user previously had? And the reason they are separate is because from `RoleHistory` we can't tell what the source was?

Reply:
Yes. I looked into whether I could differentiate admin_roles in role_history in rebuildRoles and I dont think thats possible without modifying vbSequence